cmd/kubeadm/app/phases/controlplane/volumes.go

```diff
@@ -86,37 +86,53 @@ func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) c
 		mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, caCertsPkiVolumeName, caCertsPkiVolumePath, caCertsPkiVolumePath, true, &hostPathDirectoryOrCreate)
 	}
 
+	// Merge user defined mounts and ensure unique volume and volume mount
+	// names
+	mounts.AddExtraHostPathMounts(kubeadmconstants.KubeAPIServer, cfg.APIServerExtraVolumes, true, &hostPathDirectoryOrCreate)
+	mounts.AddExtraHostPathMounts(kubeadmconstants.KubeControllerManager, cfg.ControllerManagerExtraVolumes, true, &hostPathDirectoryOrCreate)
+	mounts.AddExtraHostPathMounts(kubeadmconstants.KubeScheduler, cfg.SchedulerExtraVolumes, true, &hostPathDirectoryOrCreate)
+
 	return mounts
 }
 
 // controlPlaneHostPathMounts is a helper struct for handling all the control plane's hostPath mounts in an easy way
 type controlPlaneHostPathMounts struct {
-	volumes      map[string][]v1.Volume
-	volumeMounts map[string][]v1.VolumeMount
+	volumes      map[string]map[string]v1.Volume
+	volumeMounts map[string]map[string]v1.VolumeMount
 }
 
 func newControlPlaneHostPathMounts() controlPlaneHostPathMounts {
 	return controlPlaneHostPathMounts{
-		volumes:      map[string][]v1.Volume{},
-		volumeMounts: map[string][]v1.VolumeMount{},
+		volumes:      map[string]map[string]v1.Volume{},
+		volumeMounts: map[string]map[string]v1.VolumeMount{},
 	}
 }
 
 func (c *controlPlaneHostPathMounts) NewHostPathMount(component, mountName, hostPath, containerPath string, readOnly bool, hostPathType *v1.HostPathType) {
-	c.volumes[component] = append(c.volumes[component], staticpodutil.NewVolume(mountName, hostPath, hostPathType))
-	c.volumeMounts[component] = append(c.volumeMounts[component], staticpodutil.NewVolumeMount(mountName, containerPath, readOnly))
+	c.volumes[component][mountName] = staticpodutil.NewVolume(mountName, hostPath, hostPathType)
+	c.volumeMounts[component][mountName] = staticpodutil.NewVolumeMount(mountName, containerPath, readOnly)
 }
 
 func (c *controlPlaneHostPathMounts) AddHostPathMounts(component string, vols []v1.Volume, volMounts []v1.VolumeMount) {
-	c.volumes[component] = append(c.volumes[component], vols...)
-	c.volumeMounts[component] = append(c.volumeMounts[component], volMounts...)
+	for _, v := range vols {
+		c.volumes[component][v.Name] = v
+	}
+	for _, v := range volMounts {
+		c.volumeMounts[component][v.Name] = v
+	}
+}
+
+func (c *controlPlaneHostPathMounts) AddExtraHostPathMounts(component string, extraVols []kubeadmapi.HostPathMount, readOnly bool, hostPathType *v1.HostPathType) {
```

Comment:
hmm, please add comments to all exported funcs in this file.

Reply:
also, it's intended that extravolumes take precedence over defaults,  right? Then comment that and add unit tests thanks